pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -24,130 +24,186 @@ import (
 )
 
 const PluginClientTimeout = 45 * time.Second
 
-// draPluginClient encapsulates all dra plugin methods.
-type draPluginClient struct {
-	pluginName string
-	plugin     *Plugin
+type (
+	nodeResourceManager interface {
+		Prepare(context.Context, *grpc.ClientConn, *plugin, *drapb.NodePrepareResourcesRequest) (*drapb.NodePrepareResourcesResponse, error)
+		Unprepare(context.Context, *grpc.ClientConn, *plugin, *drapb.NodeUnprepareResourcesRequest) (*drapb.NodeUnprepareResourcesResponse, error)
+	}
+
+	v1alpha2NodeResourceManager struct{}
+	v1alpha3NodeResourceManager struct{}
+)
+
+var nodeResourceManagers = map[string]nodeResourceManager{
+	v1alpha2Version: v1alpha2NodeResourceManager{},
+	v1alpha3Version: v1alpha3NodeResourceManager{},
+}
+
+func (v1alpha2rm v1alpha2NodeResourceManager) Prepare(ctx context.Context, conn *grpc.ClientConn, _ *plugin, req *drapb.NodePrepareResourcesRequest) (*drapb.NodePrepareResourcesResponse, error) {
+	nodeClient := drapbv1alpha2.NewNodeClient(conn)
+	response := &drapb.NodePrepareResourcesResponse{
+		Claims: make(map[string]*drapb.NodePrepareResourceResponse),
+	}
+
+	for _, claim := range req.Claims {
+		res, err := nodeClient.NodePrepareResource(ctx,
+			&drapbv1alpha2.NodePrepareResourceRequest{
+				Namespace:      claim.Namespace,
+				ClaimUid:       claim.Uid,
+				ClaimName:      claim.Name,
+				ResourceHandle: claim.ResourceHandle,
+			})
+		result := &drapb.NodePrepareResourceResponse{}
+		if err != nil {
+			result.Error = err.Error()
+		} else {
+			result.CDIDevices = res.CdiDevices
+		}
+		response.Claims[claim.Uid] = result
+	}
+
+	return response, nil
+}
+
+func (v1alpha2rm v1alpha2NodeResourceManager) Unprepare(ctx context.Context, conn *grpc.ClientConn, _ *plugin, req *drapb.NodeUnprepareResourcesRequest) (*drapb.NodeUnprepareResourcesResponse, error) {
+	nodeClient := drapbv1alpha2.NewNodeClient(conn)
+	response := &drapb.NodeUnprepareResourcesResponse{
+		Claims: make(map[string]*drapb.NodeUnprepareResourceResponse),
+	}
+
+	for _, claim := range req.Claims {
+		_, err := nodeClient.NodeUnprepareResource(ctx,
+			&drapbv1alpha2.NodeUnprepareResourceRequest{
+				Namespace:      claim.Namespace,
+				ClaimUid:       claim.Uid,
+				ClaimName:      claim.Name,
+				ResourceHandle: claim.ResourceHandle,
+			})
+		result := &drapb.NodeUnprepareResourceResponse{}
+		if err != nil {
+			result.Error = err.Error()
+		}
+		response.Claims[claim.Uid] = result
+	}
+
+	return response, nil
+}
+
+func (v1alpha3rm v1alpha3NodeResourceManager) Prepare(ctx context.Context, conn *grpc.ClientConn, p *plugin, req *drapb.NodePrepareResourcesRequest) (*drapb.NodePrepareResourcesResponse, error) {
+	nodeClient := drapb.NewNodeClient(conn)
+	response, err := nodeClient.NodePrepareResources(ctx, req)
+	if err != nil {
+		status, _ := grpcstatus.FromError(err)
+		if status.Code() == grpccodes.Unimplemented {
+			p.Lock()
+			p.version = v1alpha2Version
+			p.Unlock()
```

Comment:
Can we add `SetVersion()` and `GetVersion()` calls to avoid needing to make the lock calls here (and in similar places) directly.

Reply:
Sure thing, done in 55e3662b72d. I made setVersion to accept a parameter just in case of supporting future api versions.